tagformat/caseconversion/case_conversion.go

```diff
@@ -26,6 +27,231 @@ type EncodeCasingFunc func(DecodedIdentifier) string
+	if !token.IsIdentifier(s) {
+		return nil, fmt.Errorf("only characters of the Letter category or '_' can appear in strings")
+	}
+	return g.decodeGoCamelCase(s, func(r rune) bool {
+		return r == '_'
+	})
+}
+
+// DecodeGoTags decodes CamelCase, snake_case, and kebab-case strings with fully
+// capitalized acronyms into a slice of lower cased strings.
+func (g *GoCaseConverter) DecodeGoTags(s string) (DecodedIdentifier, error) {
+	return g.decodeGoCamelCase(s, func(r rune) bool {
+		return r == '_' || r == '-'
+	})
+}
+
+// decodeGoCamelCase splits up a string in a slice of lower cased sub-string by
+// splitting after fully capitalized acronyms and after the characters that
+// signal word boundaries as specified in the passed isWordBoundary function
+func (g *GoCaseConverter) decodeGoCamelCase(s string, isWordBoundary func(rune) bool) (DecodedIdentifier, error) {
+	words := []string{}
+
+	buf := strings.Builder{}
+
+	sRunes := []rune(s)
+	for i := 0; i < len(sRunes); i++ {
+		char := sRunes[i]
+		if buf.Len() > 0 && (firstCharOfInitialism(s, i) || firstCharAfterInitialism(s, i) || isWordBoundary(char)) {
+			// We think we're at a word boundary, but we need to check if this is a prefix for an atom.
+			// We're looking for the longest matching atom, so this requires some iteration.
+			offset := sort.SearchStrings(g.atoms, buf.String())
+			bestMatch := -1
+			bestMatchLenDiff := 0
+			for ; offset < len(g.atoms) && strings.HasPrefix(g.atoms[offset], buf.String()); offset++ {
+				str := buf.String()
+				candidate := g.atoms[offset]
+
+				lenDiff := len(candidate) - len(str)
+				if lenDiff > 0 && len(str) < (i+lenDiff) {
+					// pull off more characters to match the length of the atom
+					str += string(sRunes[i : i+lenDiff])
+				}
+
+				if strings.EqualFold(candidate, str) {
+					// we found an atom that matches exactly, so we should hold
+					// on to that before we look for something potentially
+					// better
+					bestMatch = offset
+					bestMatchLenDiff = lenDiff
+				}
+			}
+
+			if bestMatch >= 0 {
+				// we found a match with an atom, so advance the pointer
+				words = append(words, g.atoms[bestMatch])
+				buf.Reset()
+				i += bestMatchLenDiff - 1
+				continue
+			}
+
+			words = append(words, buf.String())
+			buf.Reset()
+
+			if isWordBoundary(char) {
+				// if we're on a word boundary, just advance past it
+				continue
+			}
+		}
+		buf.WriteRune(char)
+	}
+
+	if buf.Len() > 0 {
+		// write whatever is left over in the buffer
+		words = append(words, buf.String())
+	}
+
+	lowerCased := make([]string, 0, len(words))
+
+	// see if any of the initialisms are actually a combination of two ("JSONAPI" or something...)
+	for _, word := range words {
+		if strings.ToUpper(word) != word {
+			// it's not an initialism because it's not all uppercase
+			lowerCased = append(lowerCased, strings.ToLower(word))
+			continue
+		}
+		chunkLen := 2
+		for chunkLen <= len(word) {
```

Comment:
You shouldn't need to iterate over sizes like this.

The [doc](https://pkg.go.dev/sort#SearchStrings) for `SearchStrings` says:
> The return value is the index to insert x if x is not present

... so the longest prefix present will be a nearby predecessor to the index that's actually returned.
I think it would be better to iterate backwards from the returned index until either the offset is either no longer a prefix or you hit the beginning of the list. (assuming that it isn't an exact match)

(If we're going to do exact matches after case-conversion, we might as well use a set, and get O(1) checks for the inner loop.)

Reply:
That's a good point -- I think the longest prefix (or an exact match) will necessarily be the entry in the position at one less than the return value of `SearchStrings` or 0 if there are no prefixes.  I've updated the code to use this logic and it appears to work as expected with the fewest loop iterations.

Reply:
At some point I convinced myself that the longest prefix would be in the previous position, but while writing that comment, I had trouble convincing myself of it again (which is why I was thinking about iterating backwards)

e.g. if we're looking for `abcd` and there are entries, `a` and `aa`, then `aa` will be in the preceding position, but we need to grab `a`, which would be another position earlier.

If we enforce that entries aren't prefixes of one another, then that simplification works, though. (I think I originally thought about this scheme in the context of Fresnel's topic prefix ACLs, where it makes no sense for them to be prefixes of eachother (but, in hindsight we probably aren't validating that 🫤 ))